components/raftstore-v2/src/router/response_channel.rs

```diff
@@ -0,0 +1,448 @@
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => return Poll::Pending,
+                Err(v) => e = v,
+            };
+            if check_bit(e, set_bit).is_some() {
+                unsafe {
+                    return Poll::Ready((*self.core.res.get()).take());
+                }
+            }
+        }
+    }
+}
+
+pub struct WriteSubscriber {
+    core: Arc<EventCore<RaftCmdResponse>>,
+}
+
+impl WriteSubscriber {
+    pub async fn wait_proposed(&mut self) -> bool {
+        WaitEvent {
+            event: WriteChannel::PROPOSED_EVENT,
+            core: &self.core,
+        }
+        .await
+    }
+
+    pub async fn wait_committed(&mut self) -> bool {
+        WaitEvent {
+            event: WriteChannel::COMMITTED_EVENT,
+            core: &self.core,
+        }
+        .await
+    }
+
+    pub async fn result(&mut self) -> Option<RaftCmdResponse> {
+        WaitResult { core: &self.core }.await
+    }
+}
+
+unsafe impl Send for WriteSubscriber {}
+unsafe impl Sync for WriteSubscriber {}
+
+pub struct WriteChannel {
+    core: ManuallyDrop<Arc<EventCore<RaftCmdResponse>>>,
+}
+
+impl WriteChannel {
+    const PROPOSED_EVENT: u64 = 2;
+    const COMMITTED_EVENT: u64 = 3;
+
+    #[inline]
+    pub fn pair() -> (Self, WriteSubscriber) {
+        let core = Arc::new(EventCore {
+            event: AtomicU64::new(0),
+            res: UnsafeCell::new(None),
+            waker: AtomicWaker::new(),
+        });
+        (
+            Self {
+                core: ManuallyDrop::new(core.clone()),
+            },
+            WriteSubscriber { core },
+        )
+    }
+}
+
+impl ErrorCallback for WriteChannel {
+    fn report_error(self, err: RaftCmdResponse) {
+        self.set_result(err);
+    }
+
+    fn is_none(&self) -> bool {
+        false
+    }
+}
+
+impl WriteCallback for WriteChannel {
+    type Response = RaftCmdResponse;
+
+    /// Called after a request is proposed to the raft group successfully. It's
+    /// used to notify the caller to move on early because it's very likely the
+    /// request will be applied to the raftstore.
+    #[inline]
+    fn notify_proposed(&mut self) {
+        self.core.notify_event(Self::PROPOSED_EVENT);
+    }
+
+    /// Called after a request is committed and before it's being applied, and
+    /// it's guaranteed that the request will be successfully applied soon.
+    #[inline]
+    fn notify_committed(&mut self) {
+        self.core.notify_event(Self::COMMITTED_EVENT);
+    }
+
+    fn trackers(&self) -> Option<&SmallVec<[TimeTracker; 4]>> {
+        None
+    }
+
+    fn trackers_mut(&mut self) -> Option<&mut SmallVec<[TimeTracker; 4]>> {
+        None
+    }
+
+    // TODO: support executing hooks inside setting result.
+    #[inline]
+    fn set_result(mut self, res: RaftCmdResponse) {
+        self.core.set_result(res);
+        unsafe {
+            ManuallyDrop::drop(&mut self.core);
+        }
+        mem::forget(self);
```

Comment:
though self.core is handled by ManuallyDrop, will the self be leaked forever?

Reply:
All allocations and drop methods are called. There is no leak.